presto-native-execution/presto_cpp/main/SystemConnector.cpp

```diff
@@ -0,0 +1,297 @@
+      velox::VARCHAR(),   velox::VARCHAR(),   velox::VARCHAR(),
+      velox::VARCHAR(),   velox::VARCHAR(),   velox::VARCHAR(),
+      velox::BIGINT(),    velox::BIGINT(),    velox::BIGINT(),
+      velox::BIGINT(),    velox::BIGINT(),    velox::BIGINT(),
+      velox::BIGINT(),    velox::BIGINT(),    velox::BIGINT(),
+      velox::BIGINT(),    velox::BIGINT(),    velox::BIGINT(),
+      velox::BIGINT(),    velox::BIGINT(),    velox::TIMESTAMP(),
+      velox::TIMESTAMP(), velox::TIMESTAMP(), velox::TIMESTAMP()};
+  kTaskSchema_ = ROW(std::move(kTaskColumnNames), std::move(kTaskColumnTypes));
+}
+
+std::string SystemTableHandle::toString() const {
+  return fmt::format("schema: {} table: {}", schemaName_, tableName_);
+}
+
+SystemDataSource::SystemDataSource(
+    const std::shared_ptr<const RowType>& outputType,
+    const std::shared_ptr<connector::ConnectorTableHandle>& tableHandle,
+    const std::unordered_map<
+        std::string,
+        std::shared_ptr<connector::ColumnHandle>>& columnHandles,
+    const TaskManager* taskManager,
+    velox::memory::MemoryPool* FOLLY_NONNULL pool)
+    : taskManager_(taskManager), pool_(pool) {
+  auto systemTableHandle =
+      std::dynamic_pointer_cast<SystemTableHandle>(tableHandle);
+  VELOX_CHECK_NOT_NULL(
+      systemTableHandle,
+      "TableHandle must be an instance of SystemTableHandle");
+  VELOX_USER_CHECK_EQ(
+      systemTableHandle->schemaName(),
+      "runtime",
+      "SystemConnector supports only runtime schema");
+  VELOX_USER_CHECK_EQ(
+      systemTableHandle->tableName(),
+      "tasks",
+      "SystemConnector supports only tasks table");
+
+  taskTableHandle_ = systemTableHandle;
+  outputColumnMappings_.reserve(outputType->names().size());
+  auto taskSchema = taskTableHandle_->taskSchema();
+  for (const auto& outputName : outputType->names()) {
+    auto it = columnHandles.find(outputName);
+    VELOX_CHECK(
+        it != columnHandles.end(),
+        "ColumnHandle is missing for output column '{}'",
+        outputName);
+
+    auto handle = std::dynamic_pointer_cast<SystemColumnHandle>(it->second);
+    VELOX_CHECK_NOT_NULL(
+        handle,
+        "ColumnHandle must be an instance of SystemColumnHandle "
+        "for '{}' on table '{}'",
+        handle->name());
+
+    auto columnIndex = taskSchema->getChildIdxIfExists(handle->name());
+    VELOX_CHECK(columnIndex.has_value());
+    outputColumnMappings_.push_back(columnIndex.value());
+  }
+
+  outputType_ = outputType;
+  taskTableResult_ = std::dynamic_pointer_cast<RowVector>(
+      BaseVector::create(taskSchema, 0, pool_));
+}
+
+void SystemDataSource::addSplit(
+    std::shared_ptr<connector::ConnectorSplit> split) {
+  VELOX_CHECK_NULL(
+      currentSplit_,
+      "Previous split has not been processed yet. Call next() to process the split.");
+  currentSplit_ = std::dynamic_pointer_cast<SystemSplit>(split);
+  VELOX_CHECK(currentSplit_, "Wrong type of split for SystemDataSource.");
+  VELOX_USER_CHECK_EQ(
```

Comment:
duplicate checks?

Reply:
I was being conservative but these can be removed.